File: windows/tests/FriendlyTerminal.Core.Tests/AgentRegistryTests.cs
using FriendlyTerminal.Core.Agents;
using Xunit;

namespace FriendlyTerminal.Core.Tests;

public class AgentRegistryTests
{
    [Theory]
    [InlineData("claude", "claude")]
    [InlineData("codex", "codex")]
    [InlineData("gemini", "gemini")]
    [InlineData("qwen", "qwen")]
    [InlineData("zcode", "zcode")]
    [InlineData("zai", "zcode")]
    public void Match_recognizes_each_tool_basename(string command, string expectedId)
    {
        Assert.Equal(expectedId, AgentRegistry.Match(command)?.Id);
    }

    [Theory]
    [InlineData("& \"C:\\Users\\me\\AppData\\Roaming\\npm\\claude.cmd\"", "claude")]
    [InlineData("/usr/local/bin/codex", "codex")]
    [InlineData("C:\\tools\\gemini.exe --resume", "gemini")]
    [InlineData("& 'C:\\Users\\John Smith\\AppData\\Roaming\\npm\\claude.cmd'", "claude")]
    [InlineData("& \"C:\\Program Files\\Codex\\codex.exe\" --resume", "codex")]
    [InlineData("'/Users/x/My Tools/claude' --continue", "claude")]
    public void Match_recognizes_path_prefixed_basename(string command, string expectedId)
    {
        Assert.Equal(expectedId, AgentRegistry.Match(command)?.Id);
    }

    [Theory]
    [InlineData("echo hi | claude", "claude")]
    [InlineData("sudo claude", "claude")]
    [InlineData("command codex", "codex")]
    [InlineData("time gemini", "gemini")]
    [InlineData("FOO=bar claude", "claude")]
    [InlineData("env QWEN_KEY=1 qwen", "qwen")]
    public void Match_skips_pipe_wrappers_and_env_assignments(string command, string expectedId)
    {
        Assert.Equal(expectedId, AgentRegistry.Match(command)?.Id);
    }

    [Theory]
    [InlineData("")]
    [InlineData("ls -la")]
    [InlineData("vim notes.txt")]
    [InlineData("git commit")]
    [InlineData("npm install claude")]
    public void Match_returns_null_for_unknown_commands(string command)
    {
        Assert.Null(AgentRegistry.Match(command));
    }

    [Fact]
    public void DangerFlags_are_matched_case_sensitively_against_the_command()
    {
        var claude = AgentRegistry.Claude;
        Assert.Contains("--dangerously-skip-permissions", claude.DangerFlags);

        var codex = AgentRegistry.Codex;
        Assert.Contains("--yolo", codex.DangerFlags);
        Assert.Empty(AgentRegistry.Zcode.DangerFlags);
    }
}

File: macosTests/AgentRegistryTests.swift
import XCTest
@testable import FriendlyTerminal

final class AgentRegistryTests: XCTestCase {

    func testEachToolBasenameMatches() {
        XCTAssertEqual(AgentRegistry.match("claude")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("codex")?.id, "codex")
        XCTAssertEqual(AgentRegistry.match("gemini")?.id, "gemini")
        XCTAssertEqual(AgentRegistry.match("qwen")?.id, "qwen")
        XCTAssertEqual(AgentRegistry.match("zcode")?.id, "zcode")
        XCTAssertEqual(AgentRegistry.match("zai")?.id, "zcode")
    }

    func testMatchIsCaseInsensitive() {
        XCTAssertEqual(AgentRegistry.match("CLAUDE")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("Gemini")?.id, "gemini")
    }

    func testPathPrefixedBasenameMatches() {
        XCTAssertEqual(AgentRegistry.match("/usr/local/bin/claude")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("/opt/homebrew/bin/codex --version")?.id, "codex")
        XCTAssertEqual(AgentRegistry.match("~/.local/bin/qwen")?.id, "qwen")
    }

    func testArgumentsDoNotAffectMatch() {
        XCTAssertEqual(AgentRegistry.match("claude --dangerously-skip-permissions")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("codex --yolo")?.id, "codex")
        XCTAssertEqual(AgentRegistry.match("gemini -y")?.id, "gemini")
    }

    func testQuotedPathsWithSpacesMatch() {
        XCTAssertEqual(AgentRegistry.match("'/Users/x/My Tools/claude'")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("\"/Users/x/My Tools/claude\" --resume")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("sudo '/opt/some dir/codex' --yolo")?.id, "codex")
        XCTAssertNil(AgentRegistry.match("'/Users/x/My Tools/claudia'"))
    }

    func testPipeUsesLastStage() {
        XCTAssertEqual(AgentRegistry.match("echo hi | claude")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("cat prompt.txt | codex")?.id, "codex")
        XCTAssertNil(AgentRegistry.match("claude | grep foo"))
    }

    func testSkipsEnvAssignmentsAndWrappers() {
        XCTAssertEqual(AgentRegistry.match("FOO=bar claude")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("sudo claude")?.id, "claude")
        XCTAssertEqual(AgentRegistry.match("command gemini")?.id, "gemini")
        XCTAssertEqual(AgentRegistry.match("exec qwen")?.id, "qwen")
        XCTAssertEqual(AgentRegistry.match("time codex")?.id, "codex")
        XCTAssertEqual(AgentRegistry.match("env FOO=bar sudo claude --resume")?.id, "claude")
    }

    func testUnknownReturnsNil() {
        XCTAssertNil(AgentRegistry.match("vim"))
        XCTAssertNil(AgentRegistry.match("ls -la"))
        XCTAssertNil(AgentRegistry.match(""))
        XCTAssertNil(AgentRegistry.match("FOO=bar"))
        XCTAssertNil(AgentRegistry.match("claudia"))
    }

    func testDangerFlagsPerProfile() {
        XCTAssertEqual(AgentRegistry.claude.dangerFlags, ["--dangerously-skip-permissions"])
        XCTAssertTrue(AgentRegistry.codex.dangerFlags.contains("--yolo"))
        XCTAssertTrue(AgentRegistry.zcode.dangerFlags.isEmpty)
    }

    // Canonical strings copied from windows/src/FriendlyTerminal.Core/Agents/AgentRegistry.cs
    func testCodexSlashCommandsMatchWindowsCanon() {
        XCTAssertEqual(
            AgentRegistry.codex.slashCommands.map { [$0.label, $0.sendText, $0.help] },
            [
                ["/status", "/status\r", "Show session status"],
                ["/model", "/model\r", "Switch the model"],
                ["/approvals", "/approvals\r", "Change the approval policy"],
                ["/compact", "/compact\r", "Compact context to save tokens"],
                ["/diff", "/diff\r", "Show pending changes"],
                ["/new", "/new\r", "Start a new conversation"],
            ]
        )
    }

    func testBaselineProfileStaysMinimal() {
        XCTAssertEqual(AgentRegistry.zcode.slashCommands.map(\.label), ["/help"])
        XCTAssertFalse(AgentRegistry.zcode.supportsNumberedOptions)
    }
}
